oneandone/src/main/java/org/apache/jclouds/oneandone/rest/compute/OneandoneComputeServiceAdapter.java

```diff
@@ -179,9 +220,54 @@
       return new NodeAndInitialCredentials<Server>(updateServer, updateServer.id(), serverCredentials);
    }
 
+   private HardwareFlavour isFlavor(String hardwareId) {
+      List<HardwareFlavour> availableProfiles = listHardwareProfiles();
+      for (HardwareFlavour flav : availableProfiles) {
+         if (flav.name() == null ? hardwareId == null : flav.name().equals(hardwareId)) {
+            return flav;
+         }
+      }
+      return null;
+   }
+
+   private ServerAppliance findWorkingImage(String imageId, Types.ServerTypeCompatibility serverType) {
+      //check if image is of Type IMAGE
+      SingleServerAppliance appliance = api.serverApplianceApi().get(imageId);
+      if (appliance.type() != Types.ApplianceType.IMAGE) {
+         //find a same OS appliance of the correct type
+         GenericQueryOptions ops = new GenericQueryOptions();
+         ops.options(0, 0, "", appliance.osVersion(), "");
+         List<ServerAppliance> allImages = api.serverApplianceApi().list(ops);
+         for (ServerAppliance app : allImages) {
+            if (app.type() == Types.ApplianceType.IMAGE
+                    && (app.osVersion() == null ? appliance.osVersion() == null : app.osVersion().equals(appliance.osVersion()))
+                    && app.osArchitecture() == appliance.osArchitecture()
+                    && app.serverTypeCompatibility().contains(serverType)
+                    && app.osImageType() == Types.OSImageType.STANDARD) {
+               return app;
+            }
+         }
+      }
+      return null;
+   }
+
    @Override
    public List<HardwareFlavour> listHardwareProfiles() {
-      return api.serverApi().listHardwareFlavours();
+      List<HardwareFlavour> result = new ArrayList<HardwareFlavour>();
+      List<HardwareFlavour> cloudModels = api.serverApi().listHardwareFlavours();
+      for (HardwareFlavour flavor : cloudModels) {
+         result.add(flavor);
+      }
+      List<BareMetalModel> baremetalModels = api.serverApi().listBaremetalModels();
+      for (BareMetalModel model : baremetalModels) {
```

Comment:
maybe you can consider a `Function<BareMetalModel, HardwareFlavour>` for better readability and testability?

Reply:
I need to combine `BareMetalModel` with `HardwareFlavour` as one list, to mix the baremetal flavors with the cloud flavors, do you think that could be achieved using the function you suggested?